crates/merkle-inclusion/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+        if let Some(b) = bit.get_value() {
+            if b {
+                actual_root_hash = GD::digest(
+                    cs.namespace(|| format!("sibling {}", i)),
+                    &[sibling_hash.to_owned(), actual_root_hash].concat(),
+                )?
+            } else {
+                actual_root_hash = GD::digest(
+                    cs.namespace(|| format!("sibling {}", i)),
+                    &[actual_root_hash, sibling_hash.to_owned()].concat(),
+                )?
+            }
+        } else {
+            return Err(SynthesisError::Unsatisfiable);
+        }
+    }
+
+    hash_equality(cs, expected_root, actual_root_hash)
+}
+
+/// Compares two hash values for equality bit by bit.
+///
+/// # Arguments
+/// * `cs` - A mutable reference to the constraint system.
+/// * `expected` - The expected hash value.
+/// * `actual` - The actual hash value to compare against.
+///
+/// # Returns
+/// A result containing the actual hash value if the hashes are equal, or a `SynthesisError` otherwise.
+fn hash_equality<E, CS>(
+    mut cs: CS,
+    expected: HashValue,
+    actual: HashValue,
+) -> Result<HashValue, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    if expected.len() != actual.len() {
+        return Err(SynthesisError::Unsatisfiable);
+    }
+
+    // Check if the reconstructed root hash matches the expected root hash
+    for i in 0..expected.len() {
+        Boolean::enforce_equal(
+            cs.namespace(|| format!("equality on {} hash bit", i)),
+            &expected[i],
+            &actual[i],
+        )?;
+    }
+
+    Ok(actual)
+}
+
+/// Conditionally reverses the order of elements in an iterator based on the endianness specified in [`GadgetDigest`].
+///
+/// This function is used to ensure that the data is correctly aligned for the hashing process in cryptographic circuits.
+/// It takes an iterator over `Boolean` values and reverses it if the [`GadgetDigest`] associated with the provided field
+/// `E` indicates little-endian order.
+///
+/// # Type Parameters
+///
+/// * `I`: A type that implements [`DoubleEndedIterator`] over references to [`Boolean`].
+/// * `E`: The prime field associated with the [`GadgetDigest`].
+/// * `GD`: The [`GadgetDigest`] implementation, which includes an endianness indicator.
+///
+/// # Arguments
+///
+/// * `iter`: An iterator over references to [`Boolean`]. This iterator is double-ended, allowing for efficient reversal
+/// if needed.
+///
+/// # Returns
+///
+/// Returns a boxed double-ended iterator over references to [`Boolean`]. The order of elements in the iterator will be
+/// reversed if [`GadgetDigest`] indicates little-endian order.
+fn conditional_reverse<
+    'a,
+    I: DoubleEndedIterator<Item = &'a Boolean> + 'a,
+    E: PrimeField,
+    GD: GadgetDigest<E>,
+>(
+    iter: I,
+) -> Box<dyn DoubleEndedIterator<Item = &'a Boolean> + 'a> {
```

Comment:
It would be great if we could inline this and avoid the performance cost of `Box<dyn MyType>` for these few lines.